Checker on GAE

k2wanko opened the pull request:
- It works with GAE.
- Add SlackWebhookNotifier

please review.

k2wanko committed:
Move main.sources to nschecker.Sources

k2wanko committed:
Can run it with GAE

k2wanko committed:
Add GAE doc

k2wanko committed:
Add SlackWebhook

haya14busa commented on backend/server.go:
```diff
@@ -0,0 +1,84 @@
+		wg.Add(1)
+		go func(s nschecker.Source) {
+			defer wg.Done()
+			err := check(ctx, hc, s)
```
[nit] `if err := check(ctx, hc, s); err != nil {`

haya14busa commented on backend/server.go:
```diff
@@ -0,0 +1,84 @@
+		if channel == "" {
+			return nil, errors.New("Please set enviroment variable SLACK_CHANNEL")
+		}
+
```
[nit] needless blank line?

haya14busa commented on cmd/nintendo-switch-checker/main.go:
```diff
@@ -221,6 +79,19 @@ func main() {
+		}
+
+		if *channel == "" {
+			log.Println("Please set -slack-channel flag")
```
s/-slack-channel/-channel/ or please change cli flag

haya14busa commented on nschecker/checker.go:
```diff
@@ -40,8 +40,27 @@ func (s State) String() string {
+	}
+}
+
+func Check(s Source, opts ...CheckOption) (State, error) {
```
hmm... i'm not sure it's good interface...

I think just passing HTTP.Client is fine. Use DefaultClient if it's nil.

haya14busa commented on nschecker/notify.go:
```diff
@@ -20,18 +18,19 @@ type Notifier interface {
 
 // SlackNotifier struct is construct an slack message.
```
I didn't notice it before but "is construct" is not valid English, i guess.

"SlackNotifier handles notification to slack" or something is better.

haya14busa commented on nschecker/notify.go:
```diff
@@ -9,8 +9,6 @@ import (
-	"github.com/nlopes/slack"
 )
 
 // Notifier interface is construct an response.
```
it does not only construct an response.

It creates request and send them  to notification service.

```
// Notifier is interface of  notification
```
or something is fine

haya14busa commented on nschecker/notify.go:
```diff
@@ -58,8 +57,24 @@ func (n *SlackNotifier) Notify(state State, s Source) error {
 	msg := fmt.Sprintf("%s%v: %v (%v)", channel, state, s.URL, s.Name)
-	params := slack.PostMessageParameters{EscapeText: false}
-	_, _, err := n.Cli.PostMessage(n.channel, msg, params)
+	v := url.Values{}
```
why did you change SlackNotifier?

Please add comment or describe the reason in commit message.

k2wanko replied:
`github.com/nlopes/slack` was not build in go1.6.
GAE/Go is only go1.6.

haya14busa commented on nschecker/notify.go:
```diff
@@ -117,3 +132,64 @@ func (n *LineNotifier) Notify(state State, s Source) error {
 	return nil
 }
+
+// SlackWebhookNotifier struct is construct an slack incoming webhook message.
```
[ditto]

haya14busa reviewed (commented):
Thanks :bird: :+1:

haya14busa commented:
I heard that some source doesn't work in GAE.
Can you describe it about this? 
If it doesn't work, I think we need to filter the source.

k2wanko committed:
Fix

- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123148511
- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123148554
- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123148890
- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123149323
- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123149092
- https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123149453

k2wanko commented:
https://github.com/haya14busa/nintendo-switch-checker/pull/3#issuecomment-309956749

@haya14busa 
I do not know the reason 😭 
will split source.

log:

```
Check failed: Yodobashi - Nintendo Switch Joy-Con(L)/(R)グレー [Nintendo Switch本体]: Get http://www.yodobashi.com/product/100000001003431565/: API error 12 (urlfetch: CONNECTION_ERROR)
Check failed: Yodobashi - Nintendo Switch Joy-Con(L)ネオンブルー/(R)ネオンレッド [Nintendo Switch本体]: Get http://www.yodobashi.com/product/100000001003431566/: API error 12 (urlfetch: CONNECTION_ERROR)
Check failed: Joshin - Nintendo Switch 本体【Joy-Con(L) ネオンブルー/(R) ネオンレッド】: Get http://joshinweb.jp/game/40519/4902370535716.html: API error 5 (urlfetch: DEADLINE_EXCEEDED)
Check failed: Joshin - Nintendo Switch 本体【Joy-Con(L)/(R) グレー】: Get http://joshinweb.jp/game/40519/4902370535709.html: API error 5 (urlfetch: DEADLINE_EXCEEDED)
```

k2wanko committed:
Split source https://github.com/haya14busa/nintendo-switch-checker/pull/3#issuecomment-309968587

k2wanko commented:
@haya14busa please review.

haya14busa commented on backend/server.go:
```diff
@@ -0,0 +1,82 @@
+	ctx := appengine.NewContext(r)
+	hc := urlfetch.Client(ctx)
+	var wg sync.WaitGroup
+	for _, s := range nschecker.Sources {
```
Please filter sources here instead of splitting the sources.

It's exceptional cases for GAE and should be handled here.

Leaving some comments also will help.

haya14busa commented:
Can you change  backend/ to appengine/, app/, server/ or something if you come up with a good term?

k2wanko committed:
rename package to server https://github.com/haya14busa/nintendo-switch-checker/pull/3#issuecomment-310062197

k2wanko commented:
@haya14busa changed

haya14busa commented:
Please address this comment. https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123229832

k2wanko committed:
Fix https://github.com/haya14busa/nintendo-switch-checker/pull/3#discussion_r123229832

k2wanko commented:
@haya14busa oops 
please check.

haya14busa reviewed (approved):
:bird: :+1:
